nixos/incus: init module and tests

adamcstephens opened the pull request:
## Description of changes

Copied, updated, and cleaned up LXD module. This one is a bit more opinionated with fewer options.

Following upstream packaging guidelines https://github.com/lxc/incus/blob/bd0ec82a9e50fb7d85afef3067685dec47600650/doc/packaging.md
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

adamcstephens committed:
nixos/incus: init module and tests

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,251 @@
+
+  options = {
+    virtualisation.incus = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = lib.mdDoc ''
+          This option enables incusd, a daemon that manages
+          containers. Users in the "incus-admin" group can interact with
+          the daemon (e.g. to start or stop containers) using the
+          {command}`incus` command line tool, among others.
+        '';
+      };
```
```suggestion
      enable = lib.mkEnableOption (lib.mdDoc ''
        incusd, a daemon that manages containers.
      
        Users in the "incus-admin" group can interact with
        the daemon (e.g. to start or stop containers) using the
        {command}`incus` command line tool, among others.
      '');
```

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,251 @@
+        '';
+      };
+
+      package = lib.mkOption {
+        type = lib.types.package;
+        default = pkgs.incus;
+        defaultText = lib.literalExpression "pkgs.incus";
+        description = lib.mdDoc ''
+          The incus package to use.
+        '';
+      };
```
```suggestion
      package = lib.mkPackageOptionMD pkgs "incus" { };
```

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,251 @@
+        '';
+      };
+
+      lxcPackage = lib.mkOption {
+        type = lib.types.package;
+        default = pkgs.lxc;
+        defaultText = lib.literalExpression "pkgs.lxc";
+        description = lib.mdDoc ''
+          The LXC package to use with Incus (required for AppArmor profiles).
+        '';
+      };
```
```suggestion
      lxcPackage = lib.mkPackageOptionMD pkgs "lxc" { };
```

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,251 @@
+      );
+
+      startTimeout = lib.mkOption {
+        type = lib.types.int;
```
```suggestion
        type = lib.types.ints.unsigned;
```

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,251 @@
+        '';
+      };
+
+      socketActivation = lib.mkEnableOption (
+        lib.mdDoc ''
+          Use socket-activation to start incus.service. Enabling this option
+          will stop the service from starting automatically on boot.
+        ''
+      );
```
Have a look at the implementation of `mkEnableOption`: https://github.com/NixOS/nixpkgs/blob/c5ada8b4ca8975d380020a085425a2cc7bbbb300/lib/options.nix#L98C5-L98C5

It automatically prepends *Whether to enable* to your string, which would make this become *Whether to enable Use socket-activation to start ....*

adamcstephens replied:
Thanks, I was unaware of this prepend.

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,251 @@
+# Systemd services for Incus.
```
This information is a bit redundant, given the context of this file.

```suggestion
```

h7x4 reviewed (changes requested):
Just a few comments about the module. Haven't looked at the tests.

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,233 @@
+      after = [ "incus.service" ];
+
+      script = ''
+        ${pkgs.coreutils}/bin/cat ${
+          preseedFormat.generate "incus-preseed.yaml" cfg.preseed
+        } | ${cfg.package}/bin/incus admin init --preseed
```
I believe the cat to stdin is unnecessary, this should do the trick.

```suggestion
        ${cfg.package}/bin/incus admin init --preseed <${
          preseedFormat.generate "incus-preseed.yaml" cfg.preseed
        }
```

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,233 @@
+    systemd.services.incus = {
+      description = "Incus Container and Virtual Machine Management Daemon";
+
+      wantedBy = lib.optional (!cfg.socketActivation) "multi-user.target";
+      after = [
+        "network-online.target"
+        "lxcfs.service"
+      ] ++ (lib.optional cfg.socketActivation "incus.socket");
+      requires = [
+        "network-online.target"
+        "lxcfs.service"
+      ] ++ (lib.optional cfg.socketActivation "incus.socket");
+
+      path = lib.optional config.boot.zfs.enabled config.boot.zfs.package;
```
Semantically speaking, it's not like we're planning to set these to the empty string if the predicates are false, it's more like we don't want to set them at all.

```suggestion
      wantedBy = lib.mkIf (!cfg.socketActivation) "multi-user.target";
      after = [
        "network-online.target"
        "lxcfs.service"
      ] ++ (lib.optional cfg.socketActivation "incus.socket");
      requires = [
        "network-online.target"
        "lxcfs.service"
      ] ++ (lib.optional cfg.socketActivation "incus.socket");

      path = lib.mkIf config.boot.zfs.enabled config.boot.zfs.package;
```

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,113 @@
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "1"
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
```
Not really a big deal, but maybe it would be better to split this up, to make it a bit easier to read? We could add error messages as well.

```suggestion
        cpuinfo = machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
        assert cpuinfo == "1", "Wrong number of CPUs reported from /proc/cpuinfo"
```

adamcstephens replied:
I've cleaned these all up. I didn't write most of this, but do agree much of this could be improved.

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,113 @@
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "2"
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
```
```suggestion
        cpuinfo = machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
        assert cpuinfo == "2", "Wrong number of CPUs reported from /proc/cpuinfo"
```

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,113 @@
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "MemTotal:          62500 kB"
+            == machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        )
```
Let's add a bit more fault tolerance, and not test the indentation

```suggestion
        meminfo = machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
        meminfo_bytes = " ".join(meminfo.split(' ')[-2:])
        assert meminfo_bytes == "62500 kB", "Wrong amount of memory reported from /proc/meminfo"
```

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,113 @@
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        assert (
+            "MemTotal:         125000 kB"
+            == machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        )
```
```suggestion
        meminfo = machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
        meminfo_bytes = " ".join(meminfo.split(' ')[-2:])
        assert meminfo_bytes == "125000 kB", "Wrong amount of memory reported from /proc/meminfo"
```

h7x4 commented on nixos/tests/incus/virtual-machine.nix:
```diff
@@ -0,0 +1,57 @@
+      # Building documentation makes the test unnecessarily take a longer time:
+      documentation.enable = lib.mkForce false;
+
+      # Our tests require `grep` & friends:
+      environment.systemPackages = with pkgs; [busybox];
```
Where does this use grep?

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,113 @@
+            == machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        )
+
+        ## ------------- ##
+        ## limits.memory ##
```
These headers seem to indicate that these assertions should be organized into different logical sections. Maybe putting them into each their own subtest would be a better way of doing this?

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,233 @@
+    };
+  };
+
+  ###### implementation
```
This comment doesn't really add anything. I've seen them in a bunch of modules, but I believe they are a relic of the past.

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,113 @@
+            == machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        )
+
+        machine.succeed("incus delete -f container")
```
If you end up making the CPU and memory section into their own subtests, just make sure this line doesn't accidentally end up as part of the memory test.

adamcstephens replied:
I didn't. There's no need to create a whole separate container for these resource checks.

h7x4 replied:
I'm not sure we're talking about the same thing. With "subtest", I mean the python `with subtest("set limits.memory"):` log grouping feature.

h7x4 replied:
@adamcstephens Hope you don't mind me tagging you here, just in case you didn't see the last comment

adamcstephens replied:
No, I don't mind at all. Maybe I wasn't clear, but I thought I understood your suggestion here.

I opted to not make the CPU and memory tests separate subtests, so it wasn't necessary to create and delete multiple containers. Or am I missing something else?

h7x4 replied:
I think the origin of my confusion is the "create and delete multiple containers" thing.

Unless I've misunderstood it completely, the subtest statement doesn't create a new test that requires you to spin up another VM, it just groups the different parts of a test under logical sections, in kind of the same way that you are grouping them with the comment titles. They even tell you which part failed if there's an error.

Please have a look at https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/test-driver/test_driver/driver.py#L88-L96

adamcstephens replied:
Right, it wouldn’t necessarily require using multiple separate containers. Sorry for conflating the two.

Do you think the test as written is ok with my latest changes?

h7x4 replied:
Regarding the node-wide resource changes? They seem fine, although I haven't run the tests myself yet

AFAIK, the only difference with the subtests would be that if the memory part or cpu part fails, instead of saying "subtest failed: Container resources are correctly applied", it could say something more specific like `subtest failed: should be able to set set limits.cpu`, followed by a stacktrace. And of course, the fact that it acts as sort of substitute for the comments.

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,113 @@
+    virtualisation = {
+      diskSize = 4096;
+
+      # Since we're testing `limits.cpu`, we've gotta have a known number of
+      # cores to lean on
+      cores = 2;
+
+      # Ditto, for `limits.memory`
+      memorySize = 512;
```
These comments doesn't make sense to me, because you actively override this number during testing. I think these comments should either be removed, or if the lines are required for other reasons, the comments should be updated to not cause anyone to think the lines could be removed.

adamcstephens replied:
These resources are for the test VM, not the guests. I've updated the comment.

h7x4 reviewed (commented):
Just a few more comments!

adamcstephens commented:
Thanks for the review @h7x4, feedback addressed along with:

- Added a test for socket-activation
- Cleaned up container test
- Switched to testing live container resource changes instead of rebooting container

adamcstephens committed:
nixos/incus: preseed should not trigger socket-activation

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,235 @@
+            networks = [
+              {
+                name = "incusbr0";
+                type = "bridge";
+                config = {
+                  "ipv4.address" = "10.0.100.1/24";
+                  "ipv4.nat" = "true";
+                };
+              }
+            ];
+            profiles = [
+              {
+                name = "default";
+                devices = {
+                  eth0 = {
+                    name = "eth0";
+                    network = "incusbr0";
+                    type = "nic";
+                  };
+                  root = {
+                    path = "/";
+                    pool = "default";
+                    size = "35GiB";
+                    type = "disk";
+                  };
+                };
+              }
+            ];
+            storage_pools = [
+              {
+                name = "default";
+                driver = "dir";
+                config = {
+                  source = "/var/lib/incus/storage-pools/default";
+                };
+              }
+            ];
+          }
+        '';
+      };
```
While this is quite a large expression, it only seems to use simple types, and the `literalExpression` can therefore be removed.

https://github.com/NixOS/nixpkgs/blob/c8e7ce71bbdb32ae51f53a6c5f6810c5e0a7e1e4/lib/options.nix#L373-L380

h7x4 commented on nixos/tests/incus/container.nix:
```diff
@@ -0,0 +1,83 @@
+        machine.succeed("echo true | incus exec container /run/current-system/sw/bin/bash -")
+        machine.succeed("incus delete -f container")
+
+    with subtest("Container resources are correctly applied"):
+        machine.succeed("incus launch nixos container")
+        with machine.nested("Waiting for instance to start and be usable"):
+            retry(instance_is_up)
+
+        machine.succeed("incus config set container limits.cpu 1")
+        # give container a chance to apply new resource limit
+        machine.sleep(1)
+        cpuinfo = machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        assert cpuinfo == "1", f"Wrong number of CPUs reported from /proc/cpuinfo, want: 1, got: {cpuinfo}"
+
+        machine.succeed("incus config set container limits.cpu 2")
+        machine.sleep(1)
+        cpuinfo = machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
+        assert cpuinfo == "2", f"Wrong number of CPUs reported from /proc/cpuinfo, want: 2, got: {cpuinfo}"
+
+        machine.succeed("incus config set container limits.memory 64MB")
+        machine.sleep(1)
+        meminfo = machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        meminfo_bytes = " ".join(meminfo.split(' ')[-2:])
+        assert meminfo_bytes == "62500 kB", f"Wrong amount of memory reported from /proc/meminfo, want: '62500 kB', got: '{meminfo_bytes}'"
+
+        machine.succeed("incus config set container limits.memory 128MB")
+        machine.sleep(1)
+        meminfo = machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
+        meminfo_bytes = " ".join(meminfo.split(' ')[-2:])
+        assert meminfo_bytes == "125000 kB", f"Wrong amount of memory reported from /proc/meminfo, want: '62500 kB', got: '{meminfo_bytes}'"
+
+        machine.succeed("incus delete -f container")
```
We already had this discussion, but I'll just leave this suggestion here.

```suggestion
    with subtest("Container is able to launch"):
        machine.succeed("incus launch nixos container")
        with machine.nested("Waiting for instance to start and be usable"):
            retry(instance_is_up)
            
    with subtest("Container CPU limits can be managed"):
        machine.succeed("incus config set container limits.cpu 1")
        # give container a chance to apply new resource limit
        machine.sleep(1)
        cpuinfo = machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
        assert cpuinfo == "1", f"Wrong number of CPUs reported from /proc/cpuinfo, want: 1, got: {cpuinfo}"

        machine.succeed("incus config set container limits.cpu 2")
        machine.sleep(1)
        cpuinfo = machine.succeed("incus exec container grep -- -c ^processor /proc/cpuinfo").strip()
        assert cpuinfo == "2", f"Wrong number of CPUs reported from /proc/cpuinfo, want: 2, got: {cpuinfo}"
        
    with subtest("Container memory limits can be managed"):
        machine.succeed("incus config set container limits.memory 64MB")
        machine.sleep(1)
        meminfo = machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
        meminfo_bytes = " ".join(meminfo.split(' ')[-2:])
        assert meminfo_bytes == "62500 kB", f"Wrong amount of memory reported from /proc/meminfo, want: '62500 kB', got: '{meminfo_bytes}'"

        machine.succeed("incus config set container limits.memory 128MB")
        machine.sleep(1)
        meminfo = machine.succeed("incus exec container grep -- MemTotal /proc/meminfo").strip()
        meminfo_bytes = " ".join(meminfo.split(' ')[-2:])
        assert meminfo_bytes == "125000 kB", f"Wrong amount of memory reported from /proc/meminfo, want: '62500 kB', got: '{meminfo_bytes}'"

    machine.succeed("incus delete -f container")
```

adamcstephens replied:
I broke these into some subtests, updated it to re-use the same container for all changes, and dropped the final deletion which isn't necessary at the end of the test.

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,235 @@
+    };
+
+    systemd.sockets.incus = lib.mkIf cfg.socketActivation {
+      description = "incus UNIX socket";
```
nit:
```suggestion
      description = "Incus UNIX socket";
```

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,235 @@
+        apply = toString;
+        description = lib.mdDoc ''
+          Time to wait (in seconds) for incusd to become ready to process requests.
+          If incusd does not reply within the configured time, incus.service will be
```
nit:

```suggestion
          If incusd does not reply within the configured time, `incus.service` will be
```

h7x4 reviewed (commented):
This is looking almost ready for merge if you ask me. Just a few last comments.

h7x4 commented on nixos/modules/virtualisation/incus.nix:
```diff
@@ -0,0 +1,235 @@
+          {
+            networks = [
+              {
+                name = "incusbr0";
+                type = "bridge";
+                config = {
+                  "ipv4.address" = "10.0.100.1/24";
+                  "ipv4.nat" = "true";
+                };
+              }
+            ];
+            profiles = [
+              {
+                name = "default";
+                devices = {
+                  eth0 = {
+                    name = "eth0";
+                    network = "incusbr0";
+                    type = "nic";
+                  };
+                  root = {
+                    path = "/";
+                    pool = "default";
+                    size = "35GiB";
+                    type = "disk";
+                  };
+                };
+              }
+            ];
+            storage_pools = [
+              {
+                name = "default";
+                driver = "dir";
+                config = {
+                  source = "/var/lib/incus/storage-pools/default";
+                };
+              }
+            ];
+          }
+        '';
```
Sorry about that. When I commented on dropping `literalExample`, I didn't mean just the function. I meant to provide a non-string example.

```suggestion
        example = {
          networks = [
            {
              name = "incusbr0";
              type = "bridge";
              config = {
                "ipv4.address" = "10.0.100.1/24";
                "ipv4.nat" = "true";
              };
            }
          ];
          profiles = [
            {
              name = "default";
              devices = {
                eth0 = {
                  name = "eth0";
                  network = "incusbr0";
                  type = "nic";
                };
                root = {
                  path = "/";
                  pool = "default";
                  size = "35GiB";
                  type = "disk";
                };
              };
            }
          ];
          storage_pools = [
            {
              name = "default";
              driver = "dir";
              config = {
                source = "/var/lib/incus/storage-pools/default";
              };
            }
          ];
        };
```

adamcstephens commented:
@ofborg eval

adamcstephens commented:
@ofborg test incus

adamcstephens commented:
@ofborg test incus

adamcstephens commented:
This is ready to go in my opinion. I'm now using this to manage Incus on two systems, both of which were migrated from LXD.

adamcstephens committed:
nixos/incus: shutdown instances on service stop

adamcstephens commented:
@ofborg test incus

h7x4 reviewed (approved):
Cool stuff! Great work :tada:

cafkafk reviewed (approved):
I skimmed the code and ran `nix-build -A nixosTests.incus`, looks okay to me.

adamcstephens commented:
ofborg test incus

adamcstephens commented:
@ofborg test incus

adamcstephens committed:
nixos/tests/incus: improve test resiliency under load

lilyinstarlight reviewed (approved):
Looks good to me now, thank you!
